server/src/main/scala/ch.epfl.scala.index.server/Server.scala

```diff
@@ -95,13 +97,56 @@ object Server {
       )
     }
 
+    val publishProcess = new PublishProcess()
+
     val route = {
       import akka.http.scaladsl._
       import server.Directives._
       import TwirlSupport._
 
+      def myAuthenticator(credentials: Option[HttpCredentials]): Authenticator[GithubCredentials] = {
+
+        case p@Provided(username) =>
+
+          credentials match {
+            case None => None
+            case Some(cred) =>
+
+              val upw = new String(new sun.misc.BASE64Decoder().decodeBuffer(cred.token()))
+              val userPass = upw.split(":")
+              val githubCredentials = GithubCredentials(userPass(0), userPass(1))
+              // todo - catch errors
+              if (publishProcess.authenticate(githubCredentials)) {
+
+                Some(githubCredentials)
+              } else {
+
+                None
+              }
+          }
+        case Missing => None
+      }
+
+
       rest.route ~
+      put {
+        path("publish" / Remaining) { path =>
+          entity(as[String]) { str =>
+            extractCredentials { credentials =>
+              authenticateBasic(realm = "scaladex Realm", myAuthenticator(credentials)) { cred =>
```

Comment:
@MasseGuillaume Sadly, i can't set a cookie or something, that we don't need to authenticate each request again. Any idea?